### PR TITLE
Meshlet LOD-compatible two-pass occlusion culling

### PR DESCRIPTION
**Pull request by JMS55**
Keeping track of explicit visibility per cluster between frames does not work with LODs, and leads to worse culling (using the final depth buffer from the previous frame is more accurate).

Instead, we need to generate a second depth pyramid after the second raster pass, and then use that in the first culling pass in the next frame to test if a cluster would have been visible last frame or not.

As part of these changes, the write_index_buffer pass has been folded into the culling pass for a large performance gain, and to avoid tracking a lot of extra state that would be needed between passes.

Prepass previous model/view stuff was adapted to work with meshlets as well.

Also fixed a bug with materials, and other misc improvements.

### CODE REVIEW

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 1]
```diff
@@ -3,36 +3,57 @@
...
 @compute
 @workgroup_size(128, 1, 1) // 128 threads per workgroup, 1 instanced meshlet per thread
-fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
-    // Fetch the instance data and check for instance culling
-    if cluster_id.x >= arrayLength(&meshlet_thread_meshlet_ids) { return; }
-    let instance_id = meshlet_thread_instance_ids[cluster_id.x];
-    if should_cull_instance(instance_id) {
-        return;
-    }
+fn cull_meshlets(@builtin(workgroup_id) workgroup_id: vec3<u32>, @builtin(num_workgroups) num_workgroups: vec3<u32>, @builtin(local_invocation_id) local_invocation_id: vec3<u32>) {
```
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:25]: This line is getting long...
```suggestion
fn cull_meshlets(
    @builtin(workgroup_id) workgroup_id: vec3<u32>,
    @builtin(num_workgroups) num_workgroups: vec3<u32>,
    @builtin(local_invocation_id) local_invocation_id: vec3<u32>,
) {
```

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 2]
```diff
@@ -3,36 +3,61 @@
...
     let model_scale = max(length(model[0]), max(length(model[1]), length(model[2])));
     let bounding_spheres = meshlet_bounding_spheres[meshlet_id];
+    var culling_bounding_sphere_center = model * vec4(bounding_spheres.self_culling.center, 1.0);
+    var culling_bounding_sphere_radius = model_scale * bounding_spheres.self_culling.radius;
+
+#ifdef MESHLET_FIRST_CULLING_PASS
+    // Frustum culling
+    // TODO: Faster method from https://vkguide.dev/docs/gpudriven/compute_culling/#frustum-culling-function
+    for (var i = 0u; i < 6u; i++) {
+        if dot(view.frustum[i], culling_bounding_sphere_center) <= -culling_bounding_sphere_radius {
```
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:57]: For consistency with the CPU frustum sphere intersection code:
```suggestion
        if dot(view.frustum[i], culling_bounding_sphere_center) + culling_bounding_sphere_radius <= 0.0 {
```

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 3]
```diff
@@ -48,61 +73,60 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
...
-    let culling_bounding_sphere_center = model * vec4(bounding_spheres.self_culling.center, 1.0);
-    let culling_bounding_sphere_radius = model_scale * bounding_spheres.self_culling.radius;
-
-    // Frustum culling
-    // TODO: Faster method from https://vkguide.dev/docs/gpudriven/compute_culling/#frustum-culling-function
-    for (var i = 0u; i < 6u; i++) {
-        if !meshlet_visible { break; }
-        meshlet_visible &= dot(view.frustum[i], culling_bounding_sphere_center) > -culling_bounding_sphere_radius;
+    let aabb = project_view_space_sphere_to_screen_space_aabb(culling_bounding_sphere_center_view_space, culling_bounding_sphere_radius);
+    let depth_pyramid_size_mip_0 = vec2<f32>(textureDimensions(depth_pyramid, 0)) * 0.5; // Halve the view-space AABB size as the depth pyramid is half the view size
```
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:90]:
```suggestion
    // Halve the view-space AABB size as the depth pyramid is half the view size
    let depth_pyramid_size_mip_0 = vec2<f32>(textureDimensions(depth_pyramid, 0)) * 0.5;
```

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl (hunk 2 above)
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:57]: I'll have to check that this doesn't hurt performance. Performance is extremely important here, as it runs for 50+ million clusters a frame :). I've micro-optimized every shader quite a bit (and still have more to go, the bounding sphere calculations have too many live registers).

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 4]
```diff
@@ -48,61 +73,60 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
...
+    let height = (aabb.w - aabb.y) * depth_pyramid_size_mip_0.y;
+    let depth_level = max(0, i32(ceil(log2(max(width, height))))); // TODO: Naga doesn't like this being a u32
+    let depth_pyramid_size = vec2<f32>(textureDimensions(depth_pyramid, depth_level));
+    let aabb_top_left = vec2<u32>(aabb.xy * depth_pyramid_size);
+
+    let depth_quad_a = textureLoad(depth_pyramid, aabb_top_left, depth_level).x;
+    let depth_quad_b = textureLoad(depth_pyramid, aabb_top_left + vec2(1u, 0u), depth_level).x;
+    let depth_quad_c = textureLoad(depth_pyramid, aabb_top_left + vec2(0u, 1u), depth_level).x;
+    let depth_quad_d = textureLoad(depth_pyramid, aabb_top_left + vec2(1u, 1u), depth_level).x;
+    let occluder_depth = min(min(depth_quad_a, depth_quad_b), min(depth_quad_c, depth_quad_d));
```
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:101]: I know it's not new code but could/shouldn't this use: https://www.w3.org/TR/WGSL/#texturegather ?

Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:102]: textureGather doesn't allow you to specify mip leve, so we can't.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 5]
```diff
@@ -48,61 +73,60 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
...
+    } else {
+        // Perspective
+        let sphere_depth = -view.projection[3][2] / (culling_bounding_sphere_center_view_space.z + culling_bounding_sphere_radius);
+        meshlet_visible = sphere_depth >= occluder_depth;
     }
 
-#ifdef MESHLET_SECOND_CULLING_PASS
-    // In the second culling pass, cull against the depth pyramid generated from the first pass
+    // Write if the cluster should be occlusion tested in the second pass
+#ifdef MESHLET_FIRST_CULLING_PASS
```
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:117]:
```suggestion
#ifdef MESHLET_FIRST_CULLING_PASS
    // Write if the cluster should be occlusion tested in the second pass
```

Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:117]: It's easier to read if we keep the comments on top of the ifdefs imo, I'd rather not change it unless you feel strongly about it (but I also want to block this PR anymore than I have to :P)

File: crates/bevy_pbr/src/meshlet/gpu_scene.rs [hunk 6]
```diff
@@ -384,10 +379,18 @@ pub fn prepare_meshlet_per_frame_resources(
                     base_mip_level: i,
                     mip_level_count: Some(1),
                     base_array_layer: 0,
-                    array_layer_count: None,
+                    array_layer_count: Some(1),
```
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/gpu_scene.rs:382]: What difference does this make?

Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/gpu_scene.rs:382]: Nothing really. A little bit more explicit and maybe saves a teeny bit of wgpu validation.

File: crates/bevy_pbr/src/prepass/mod.rs [hunk 7]
```diff
@@ -199,9 +200,14 @@ pub struct PreviousViewData {
     pub view_proj: Mat4,
 }
 
+#[cfg(not(feature = "meshlet"))]
+type PreviousViewFilter = (With<Camera3d>, With<MotionVectorPrepass>);
+#[cfg(feature = "meshlet")]
+type PreviousViewFilter = Or<(With<Camera3d>, With<ShadowView>)>;
```
Comment (superdump) [inline crates/bevy_pbr/src/prepass/mod.rs:206]: How come this one has `ShadowView`?

Comment (JMS55) [reply crates/bevy_pbr/src/prepass/mod.rs:206]: ShadowViews do not have Camera3d, but specifically for meshlets I need the previous view for them.

**Comment by JMS55** [comment]
Something is up with the bounding spheres being wonky. Occlusion culling isn't working correctly.

**Comment by JMS55** [comment]
Fixed bugs, ready to go again.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 8]
```diff
@@ -48,61 +73,59 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
...
-        if view.projection[3][3] == 1.0 {
-            // Orthographic
-            let sphere_depth = view.projection[3][2] + (culling_bounding_sphere_center_view_space.z + culling_bounding_sphere_radius) * view.projection[2][2];
-            meshlet_visible &= sphere_depth >= occluder_depth;
-        } else {
-            // Perspective
-            let sphere_depth = -view.projection[3][2] / (culling_bounding_sphere_center_view_space.z + culling_bounding_sphere_radius);
-            meshlet_visible &= sphere_depth >= occluder_depth;
+        let meshlet_triangle_count = meshlets[meshlet_id].triangle_count;
+        let buffer_start = atomicAdd(&draw_indirect_args.vertex_count, meshlet_triangle_count * 3u) / 3u;
```
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:123]: is it not possible to skip the muls and divs by 3 here and just multiply afterwards at the last possible moment?

File: crates/bevy_pbr/src/meshlet/gpu_scene.rs [hunk 9]
```diff
@@ -238,29 +244,28 @@ pub fn prepare_meshlet_per_frame_resources(
...
                     mapped_at_creation: false,
                 });
                 *slot = Some(buffer.clone());
                 buffer
             }
         };
 
-    let needed_buffer_size = gpu_scene.scene_meshlet_count.div_ceil(32) as u64 * 4;
+    let needed_buffer_size =
+        gpu_scene.scene_meshlet_count.div_ceil(32) as u64 * size_of::<u32>() as u64;
```
Comment (atlv24) [inline crates/bevy_pbr/src/meshlet/gpu_scene.rs:268]: https://doc.rust-lang.org/std/primitive.u32.html#associatedconstant.BITS
```suggestion
        gpu_scene.scene_meshlet_count.div_ceil(u32::BITS) as u64 * size_of::<u32>() as u64;
```

**Review by atlv24 — APPROVED** [review]
I still think that there's many literals in the shaders that should be either extracted into constants or conversion functions (like the `<< 6u`/`>> 6u` cluster id conversions can be functions, 32u can be BITS_PER_ELEM or similar) but this can be addressed later once those numbers actually settle

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl (hunk 8 above)
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:123]: We'd need an extra pass to just multiply by 3, I doubt it's worth it just to save like 2 instructions.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 10]
```diff
@@ -3,36 +3,61 @@
...
+    if should_cull_instance(instance_id) { return; }
+#endif
 
-    // Fetch other meshlet data
-    let meshlet_id = meshlet_thread_meshlet_ids[cluster_id.x];
+    // Calculate world-space culling bounding sphere for the cluster
     let instance_uniform = meshlet_instance_uniforms[instance_id];
+    let meshlet_id = meshlet_thread_meshlet_ids[cluster_id];
     let model = affine3_to_square(instance_uniform.model);
     let model_scale = max(length(model[0]), max(length(model[1]), length(model[2])));
```
Comment (pcwalton) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:48]:
```suggestion
    let model_scale = sqrt(max(dot(model[0], model[0]), max(dot(model[1], model[1]), dot(model[2], model[2]))));
```
Saves 2 sqrt instructions.

**Review by pcwalton — APPROVED** [review]
This looks fine. Since you mentioned performance, I left a few microoptimization suggestions. Feel free to take them or leave them :)

File: crates/bevy_pbr/src/meshlet/meshlet_bindings.wgsl [hunk 11]
```diff
@@ -56,46 +57,25 @@ struct DrawIndirectArgs {
...
-@group(0) @binding(2) var<storage, read> meshlet_previous_cluster_ids: array<u32>; // Per cluster (instance of a meshlet)
-@group(0) @binding(3) var<storage, read> meshlet_previous_occlusion: array<u32>; // 1 bit per cluster (instance of a meshlet), packed as a bitmask
-@group(0) @binding(4) var<storage, read> meshlets: array<Meshlet>; // Per asset meshlet
-@group(0) @binding(5) var<storage, read_write> draw_indirect_args: DrawIndirectArgs; // Single object shared between all workgroups/meshlets/triangles
-@group(0) @binding(6) var<storage, read_write> draw_index_buffer: array<u32>; // Single object shared between all workgroups/meshlets/triangles
-
-fn get_meshlet_occlusion(cluster_id: u32) -> bool {
-    let packed_occlusion = meshlet_occlusion[cluster_id / 32u];
+fn meshlet_is_second_pass_candidate(cluster_id: u32) -> bool {
+    // TODO: Does this read need to be an atomicLoad?
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/meshlet_bindings.wgsl:75]: Oh whoops, this is left over from the previous version of this PR. I can remove it.

File: crates/bevy_pbr/src/meshlet/visibility_buffer_raster_node.rs [hunk 12]
```diff
@@ -188,36 +176,23 @@ impl Node for MeshletVisibilityBufferRasterPassNode {
...
-                );
-            }
+            render_context.command_encoder().clear_buffer(
+                &meshlet_view_resources.second_pass_candidates_buffer,
+                0,
+                None,
+            );
             cull_pass(
-                "meshlet_culling_first_pass",
+                "culling_first",
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/visibility_buffer_raster_node.rs:188]: It is separated out. It's already nested under a larger meshlet debug span. The repeated use of meshlet for individual passes made it harder to read the NSight profiles.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl [hunk 13]
```diff
@@ -48,61 +73,59 @@ fn cull_meshlets(@builtin(global_invocation_id) cluster_id: vec3<u32>) {
...
-
-    // Frustum culling
-    // TODO: Faster method from https://vkguide.dev/docs/gpudriven/compute_culling/#frustum-culling-function
-    for (var i = 0u; i < 6u; i++) {
-        if !meshlet_visible { break; }
-        meshlet_visible &= dot(view.frustum[i], culling_bounding_sphere_center) > -culling_bounding_sphere_radius;
+    // Project the culling bounding sphere to view-space for occlusion culling
+#ifdef MESHLET_FIRST_CULLING_PASS
+    let previous_model = affine3_to_square(instance_uniform.previous_model);
+    let previous_model_scale = max(length(previous_model[0]), max(length(previous_model[1]), length(previous_model[2])));
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:81]: It's hard to tell. These microoptimizations don't seem to show up on profiles, as they're not the bottleneck. The bottleneck for culling is mostly thread divergence, and a teeny bit lack of registers.

**Review by superdump — APPROVED** [review]

**Pull request merged**